### ISSUE TITLE
[EPIC] Public API / RPC Tracking Issue

### ISSUE THREAD

**Issue by Wondertan** [celestiaorg/celestia-node#169]
## Current State
Right now, we have implemented an "RPC" server in #124 which can be started /stopped and provides a way to register arbitrary __HTTP__ handlers. Also, we have an unmerged #135 that adds the first message for the "RPC" server - StatusMessage.

## Problems
* As you may notice, I put "RPC" in quotes and the reason is that the current implementation is not an RPC server, but a plain HTTP server using std implementation. 
* Current architecture/vision for Services registering custom HTTP Handlers is prone to cyclic dependencies.
	* As noted [here](https://github.com/celestiaorg/celestia-node/pull/135#discussion_r735408661), `RegisterHandlers` was planned to be extended, so Services could register their own handlers. This is impossible because Services would need to have a pointer to the Node structure, while Node itself contains pointers to all the services. Unfortunately, this wouldn't compile due to mentioned cyclic dependency, which can not be avoided without hacky solutions.
* Simple HTTP server does not provide capabilities of the real RPC server. Even though RPC may and usually runs over HTTP, RPC provides a client/server framework and a protocol to remotely call arbitrary handlers. Using simple HTTP server requires us to:
	* Implement client code separately.
	* Implement a custom framework to work over websockets.
	* Come up with a custom authorization solution.
	* Basically, reinvent multiple wheels.

## Solution
> NOTE: It is not detailed due to current priorities and according to the current devnet attack plan, we won't need an RPC server, as https://github.com/celestiaorg/dalc is going to implement its own GRPC server.

The main idea is to use existing RPC implementation without reimplementing the wheel:
* https://github.com/etclabscore/go-openrpc-reflect
* https://github.com/filecoin-project/go-jsonrpc
* https://github.com/grpc/grpc-go
	* God pls no
* https://github.com/sourcegraph/jsonrpc2

> More details to come once this starts to be a priority again

## Action Items
- [x] #255
- [ ] #256
- [x] #257
- [x] #258
- [x] #259

**Comment by renaynay** [comment]
SGTM - let's use filecoin impl.

**Comment by renaynay** [comment]
@Wondertan let's use this issue as a main issue tracking for RPC in general and then break it down into smaller issues from here.

**Comment by liamsi** [comment]
> God pls no

What is so bad about grpc? 

> SGTM - let's use filecoin impl.

What is the rationale behind this and where can I learn more about filecoin's rpc implementation? There isn't even a Readme in the repo.

**Comment by jbowen93** [comment]
What are the specific requirements for the RPC server? Does it just have to do RPC?

If there aren't specific requirements that are only satisfied by Filecoin's RPC implementation I'm opposed to choosing an undocumented framework without widespread adoption. 

Specific criticisms of gRPC (and other alternatives) as well as unique benefits that Filecoin's implementation provides would be appreciated.

**Comment by renaynay** [comment]
Oh I was under the impression that we would base our RPC scaffolding around filecoin's rather than using filecoin's impl as a lib. Which is it @Wondertan , using as inspiration or importing it?

**Comment by renaynay** [comment]
Deprecating in favour of #944
